types/styled-system/index.d.ts

```diff
@@ -13,62 +13,85 @@
 export function get(obj: any, ...paths: Array<string | number>): any;
-export function themeGet(keys: string, fallback?: string): any;
 
-export function is(n: any): boolean;
-export function isObject(n: any): boolean;
-export function num(n: any): boolean;
-export function px(n: any): string;
+export type ObjectOrArray<T> = T[] | { [K: string]: T | ObjectOrArray<T> };
 
-export function createMediaQuery(n: number | string): string;
+export type Scale = ObjectOrArray<number | string>;
 
-export interface styleFn {
-    (...args: any[]): any;
-    propTypes?: string[];
-}
+export type TLengthStyledSystem = string | 0 | number;
 
-export type Scale = object | Array<string | number>;
+export type ResponsiveValue<T> = T | Array<T | null> | { [key: string]: T };
 
-export interface LowLevelStylefunctionArguments<N, S> {
-  prop: string;
-  cssProperty?: string;
-  alias?: string;
-  key?: string;
-  transformValue?: (n: N, scale?: S) => any;
-  scale?: S;
+// Preserved to support v4 shim:
+// https://github.com/styled-system/styled-system/blob/master/packages/styled-system/src/index.js#L108
+export interface LowLevelStyleFunctionArguments<N, S> {
+    prop: string;
+    cssProperty?: string;
+    alias?: string;
+    key?: string;
+    transformValue?: (n: N, scale?: S) => any;
+    scale?: S;
+    // new v5 api
+    properties?: string[];
 }
 
 export function style<N = string | number, S = Scale>(
-  // tslint:disable-next-line no-unnecessary-generics
-  args: LowLevelStylefunctionArguments<N, S>
-): { [cssProp: string]: string };
+    // tslint:disable-next-line no-unnecessary-generics
+    args: LowLevelStyleFunctionArguments<N, S>
+): {
+    [cssProp: string]: string;
+};
 
-export function compose(
-    ...funcs: Array<(...args: any[]) => any>
-): (...args: any[]) => any;
+export interface styleFn {
+    (...args: any[]): any;
+    config?: object;
+    propNames?: string[];
+    cache?: object;
+}
 
-export function mapProps(mapper: (...args: any[]) => any): (func: any) => (...props: any[]) => any;
+export interface ConfigStyle {
+    /** The CSS property to use in the returned style object (overridden by `properties` if present). */
+    property?: keyof CSS.Properties;
+    /**
+     * An array of multiple properties (e.g. `['marginLeft', 'marginRight']`) to which this style's value will be
+     * assigned (overrides `property` when present).
+     */
+    properties?: Array<keyof CSS.Properties>;
+    /** A string referencing a key in the `theme` object. */
+    scale?: string;
+    /** A fallback scale object for when there isn't one defined in the `theme` object. */
+    defaultScale?: Scale;
+    /** A function to transform the raw value based on the scale. */
+    transform?: (value: any, scale?: Scale) => any;
+}
+
+export interface Config {
+    /** Property name exposed for use in components */
+    [customStyleName: string]: ConfigStyle | boolean;
+}
+
+export function compose(...parsers: styleFn[]): styleFn;
```

Comment:
just wondering if theres any way to type that _only_ `styleFn`s should be passed to `compose`? currently regular functions without the extra properties are allowed, since [all the extra properties are optional](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/36041/files#diff-a483ed9502305df77fe429527806d00dR52). If any aren't truly optional for `styled-system` functions it'd be great to assert at least once, since `compose` will silently fail when a "regular" userland style function is passed in.

(p.s. thanks for doing this work! y'all are very fantastic)

Reply:
Let’s make this a follow-up and not let perfect become the enemy of good enough.